tikv-client/src/main/java/com/pingcap/tikv/txn/LockResolverClient.java

```diff
@@ -0,0 +1,288 @@
+import com.pingcap.tikv.kvproto.Metapb.Store;
+import com.pingcap.tikv.kvproto.TikvGrpc;
+import com.pingcap.tikv.kvproto.TikvGrpc.TikvStub;
+import com.pingcap.tikv.kvproto.TikvGrpc.TikvBlockingStub;
+import com.pingcap.tikv.operation.KVErrorHandler;
+import com.pingcap.tikv.region.RegionErrorReceiver;
+import com.pingcap.tikv.region.TiRegion;
+import com.pingcap.tikv.region.TiRegion.RegionVerID;
+import com.pingcap.tikv.util.BackOffer;
+import com.pingcap.tikv.util.TsoUtils;
+import io.grpc.ManagedChannel;
+import org.apache.log4j.Logger;
+
+import java.util.*;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.function.Supplier;
+
+import static com.pingcap.tikv.util.BackOffFunction.BackOffFuncType.BoRegionMiss;
+
+// LockResolver resolves locks and also caches resolved txn status.
+public class LockResolverClient extends AbstractGRPCClient<TikvBlockingStub, TikvStub> implements RegionErrorReceiver {
+  // ResolvedCacheSize is max number of cached txn status.
+  private static final long resolvedTxnCacheSize = 2048;
+  // By default, locks after 3000ms is considered unusual (the client created the
+  // lock might be dead). Other client may cleanup this kind of lock.
+  // For locks created recently, we will do backoff and retry.
+  private static final long defaultLockTTL = 3000;
+  private static final long maxLockTTL = 120000;
+  // ttl = ttlFactor * sqrt(writeSizeInMiB)
+  private static final long ttlFactor = 6000;
+  private static final Logger logger = Logger.getLogger(LockResolverClient.class);
+
+  private final ReadWriteLock readWriteLock;
+  // Note: Because the internal of long is same as unsigned_long
+  // and Txn id are never changed. Be careful to compare between two tso
+  // the `resolved` mapping is as {@code Map<TxnId, TxnStatus>}
+  // TxnStatus represents a txn's final status. It should be Commit or Rollback.
+  // if TxnStatus > 0, means the commit ts, otherwise abort
+  private final Map<Long, Long> resolved;
+  // the list is chain of txn for O(1) lru cache
+  private final LinkedList<Long> recentResolved;
+  private TikvBlockingStub blockingStub;
+  private TikvStub asyncStub;
+  private TiRegion region;
+
+  public LockResolverClient(TiSession session,
+                            TikvBlockingStub blockingStub,
+                            TikvStub asyncStub) {
+    super(session);
+    resolved = new HashMap<>();
+    recentResolved = new LinkedList<>();
+    readWriteLock = new ReentrantReadWriteLock();
+    this.blockingStub = blockingStub;
+    this.asyncStub = asyncStub;
+  }
+
+  private void saveResolved(long txnID, long status) {
+    try {
+      readWriteLock.writeLock().lock();
+      if (resolved.containsKey(txnID)) {
+        return ;
+      }
+
+      resolved.put(txnID, status);
+      recentResolved.addLast(txnID);
+      if (recentResolved.size() > resolvedTxnCacheSize) {
+        Long front = recentResolved.removeLast();
```

Comment:
Your recentResolved seems to be actually a stack. Why do you need a LinkedList instead?
Also, I think it should be addLast and removeFront according to TiDB's implementation.